GPU process uses same map for all resources

kkinnunen-apple opened the pull request:
#### c8076268408dbfa908a8ba48957726ee5f7c809b
<pre>
GPU process uses same map for all resources
<a href="https://bugs.webkit.org/show_bug.cgi?id=288587">https://bugs.webkit.org/show_bug.cgi?id=288587</a>
<a href="https://rdar.apple.com/145647088">rdar://145647088</a>

Reviewed by Matt Woodrow.

DisplayList::ResourceHeap, RemoteResourceCacheProxy, RemoteResourceCache
would use std::variant to put unrelated types ImageBuffer, NativeImage,
Gradient, DecomposedGlyphs, Font, FontCustomPlatformData to the same
hash map.

These types are distinct types that are never acted upon generically,
rather each have distinct insert, remove and query locations that
use the specific type. Thus it does not have any benefit to store
these in the same map.

Performance-wise, the type is the best discriminator and thus inserts,
lookups are fastest with the smallest amount of items in the maps.

Memory-wise, due to the superlinear overallocation of hashmaps, the
smallest amount if items in each map is the best approach when maps
are of non-trivial size.

Code structure-wise, having the different types in the same map causes
the antipattern of having the identifier as RenderingResourceIdentifier
instead of type-specific identifiers. It also requires error-prone and
buggy code instance counting, downcasting and result filtering code
that is just not needed at all.

Fix by storing the instances in HashSets and HashMaps that are distinct
to each type.

Implement the maps in both RemoteResourceCache and
DisplayList::ResourceHeap. These have different use-cases and are not
used with each others interfaces. Using DisplayList::ResourceHeap as the
implementation of RemoteResourceCache would leak remote resource cache
related code to DisplayList::ResourceHeap. Also
DisplayList::ResourceHeap is unneeded functionality and will be removed
in later commits.

Later commits will improve the type safety by introducing the type
specific identifiers.

* Source/WTF/wtf/WeakHashSet.h:
Add capability to store WeakRefs that come from WeakFactory.

* Source/WebCore/platform/graphics/DecomposedGlyphs.cpp:
(WebCore::DecomposedGlyphs::~DecomposedGlyphs):
* Source/WebCore/platform/graphics/DecomposedGlyphs.h:
* Source/WebCore/platform/graphics/Gradient.cpp:
(WebCore::Gradient::~Gradient):
* Source/WebCore/platform/graphics/Gradient.h:
* Source/WebCore/platform/graphics/NativeImage.cpp:
(WebCore::NativeImage::~NativeImage):
* Source/WebCore/platform/graphics/NativeImage.h:
* Source/WebCore/platform/graphics/RenderingResource.h:
(WebCore::RenderingResource::addObserver):
(WebCore::RenderingResource::~RenderingResource): Deleted.
(WebCore::RenderingResource::removeObserver): Deleted.
Remove unused function, it makes no sense for weak observers. Instead,
invalidate the observer.

* Source/WebCore/platform/graphics/displaylists/DisplayListItem.cpp:
(WebCore::DisplayList::applyFilteredImageBufferItem):
(WebCore::DisplayList::applyImageBufferItem):
(WebCore::DisplayList::applyNativeImageItem):
(WebCore::DisplayList::applyDrawGlyphs):
(WebCore::DisplayList::applyDrawDecomposedGlyphs):
* Source/WebCore/platform/graphics/displaylists/DisplayListResourceHeap.h:
(WebCore::DisplayList::ResourceHeap::add):
(WebCore::DisplayList::ResourceHeap::getImageBuffer):
(WebCore::DisplayList::ResourceHeap::getNativeImage):
(WebCore::DisplayList::ResourceHeap::getSourceImage):
(WebCore::DisplayList::ResourceHeap::getDecomposedGlyphs const):
(WebCore::DisplayList::ResourceHeap::getGradient const):
(WebCore::DisplayList::ResourceHeap::getFilter const):
(WebCore::DisplayList::ResourceHeap::getFont const):
(WebCore::DisplayList::ResourceHeap::clearAllResources):
(WebCore::DisplayList::ResourceHeap::getFontCustomPlatformData const): Deleted.
(WebCore::DisplayList::ResourceHeap::resources const): Deleted.
(WebCore::DisplayList::ResourceHeap::removeImageBuffer): Deleted.
(WebCore::DisplayList::ResourceHeap::removeRenderingResource): Deleted.
(WebCore::DisplayList::ResourceHeap::removeFont): Deleted.
(WebCore::DisplayList::ResourceHeap::removeFontCustomPlatformData): Deleted.
(WebCore::DisplayList::ResourceHeap::clearAllImageResources): Deleted.
(WebCore::DisplayList::ResourceHeap::clearAllDrawingResources): Deleted.
(WebCore::DisplayList::ResourceHeap::get const): Deleted.
(WebCore::DisplayList::ResourceHeap::remove): Deleted.
(WebCore::DisplayList::ResourceHeap::checkInvariants const): Deleted.
(): Deleted.
* Source/WebCore/platform/graphics/filters/FilterFunction.cpp:
(WebCore::FilterFunction::~FilterFunction):
* Source/WebCore/platform/graphics/filters/FilterFunction.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::releaseNativeImage):
(WebKit::RemoteRenderingBackend::releaseFont):
(WebKit::RemoteRenderingBackend::releaseFontCustomPlatformData):
(WebKit::RemoteRenderingBackend::releaseDecomposedGlyphs):
(WebKit::RemoteRenderingBackend::releaseGradient):
(WebKit::RemoteRenderingBackend::releaseFilter):
(WebKit::RemoteRenderingBackend::releaseRenderingResource): Deleted.
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteResourceCache.cpp:
(WebKit::RemoteResourceCache::cacheNativeImage):
(WebKit::RemoteResourceCache::releaseNativeImage):
(WebKit::RemoteResourceCache::cachedNativeImage const):
(WebKit::RemoteResourceCache::cacheDecomposedGlyphs):
(WebKit::RemoteResourceCache::releaseDecomposedGlyphs):
(WebKit::RemoteResourceCache::cachedDecomposedGlyphs const):
(WebKit::RemoteResourceCache::cacheGradient):
(WebKit::RemoteResourceCache::releaseGradient):
(WebKit::RemoteResourceCache::cachedGradient const):
(WebKit::RemoteResourceCache::cacheFilter):
(WebKit::RemoteResourceCache::releaseFilter):
(WebKit::RemoteResourceCache::cachedFilter const):
(WebKit::RemoteResourceCache::cacheFont):
(WebKit::RemoteResourceCache::releaseFont):
(WebKit::RemoteResourceCache::cachedFont const):
(WebKit::RemoteResourceCache::cacheFontCustomPlatformData):
(WebKit::RemoteResourceCache::releaseFontCustomPlatformData):
(WebKit::RemoteResourceCache::cachedFontCustomPlatformData const):
(WebKit::RemoteResourceCache::releaseAllResources):
(WebKit::RemoteResourceCache::releaseAllDrawingResources):
(WebKit::RemoteResourceCache::releaseAllImageResources):
(WebKit::RemoteResourceCache::releaseRenderingResource): Deleted.
* Source/WebKit/GPUProcess/graphics/RemoteResourceCache.h:
(WebKit::RemoteResourceCache::resourceHeap const): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::releaseImageBuffer):
(WebKit::RemoteRenderingBackendProxy::releaseNativeImage):
(WebKit::RemoteRenderingBackendProxy::releaseFont):
(WebKit::RemoteRenderingBackendProxy::releaseFontCustomPlatformData):
(WebKit::RemoteRenderingBackendProxy::releaseDecomposedGlyphs):
(WebKit::RemoteRenderingBackendProxy::releaseGradient):
(WebKit::RemoteRenderingBackendProxy::releaseFilter):
(WebKit::RemoteRenderingBackendProxy::releaseAllDrawingResources):
(WebKit::RemoteRenderingBackendProxy::releaseAllImageResources):
(WebKit::RemoteRenderingBackendProxy::releaseRenderingResource): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.cpp:
(WebKit::RemoteResourceCacheProxy::~RemoteResourceCacheProxy):
(WebKit::RemoteResourceCacheProxy::clear):
(WebKit::RemoteResourceCacheProxy::recordDecomposedGlyphsUse):
(WebKit::RemoteResourceCacheProxy::recordGradientUse):
(WebKit::RemoteResourceCacheProxy::recordFilterUse):
(WebKit::RemoteResourceCacheProxy::recordNativeImageUse):
(WebKit::RemoteResourceCacheProxy::willDestroyNativeImage):
(WebKit::RemoteResourceCacheProxy::willDestroyGradient):
(WebKit::RemoteResourceCacheProxy::willDestroyDecomposedGlyphs):
(WebKit::RemoteResourceCacheProxy::willDestroyFilter):
(WebKit::RemoteResourceCacheProxy::clearRenderingResources):
(WebKit::RemoteResourceCacheProxy::clearNativeImages):
(WebKit::RemoteResourceCacheProxy::finalizeRenderingUpdateForFonts):
(WebKit::RemoteResourceCacheProxy::remoteResourceCacheWasDestroyed):
(WebKit::RemoteResourceCacheProxy::releaseMemory):
(WebKit::RemoteResourceCacheProxy::releaseAllImageResources):
(WebKit::RemoteResourceCacheProxy::imagesCount const): Deleted.
(WebKit::RemoteResourceCacheProxy::cachedNativeImage const): Deleted.
(WebKit::RemoteResourceCacheProxy::releaseRenderingResource): Deleted.
(WebKit::RemoteResourceCacheProxy::clearRenderingResourceMap): Deleted.
(WebKit::RemoteResourceCacheProxy::clearNativeImageMap): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::remoteImagesCountForTesting const):

Canonical link: <a href="https://commits.webkit.org/291617@main">https://commits.webkit.org/291617@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/71fedecb29cde522195a235fbae11fcbe7485b59

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/93490 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/13056 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/2791 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/98492 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/44016 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/95540 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/13346 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/21506 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/71423 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/28808 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/96492 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/9983 "Found 1 new test failure: fast/forms/ios/focus-input-via-button-no-scaling.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/84543 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/51757 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/92985 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/9665 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/2167 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/43330 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/86200 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/79944 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/2215 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/100524 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/92156 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/20542 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/15018 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/80438 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/20794 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/80474 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/79768 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19825 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/24303 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/1646 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/13677 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/20526 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/25704 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/114806 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/20213 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/33082 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/23673 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/21954 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mattwoodrow commented on Source/WebCore/platform/graphics/displaylists/DisplayListResourceHeap.h:
```diff
@@ -122,193 +106,54 @@ class ResourceHeap {
     }
 
-    Font* getFont(RenderingResourceIdentifier renderingResourceIdentifier) const
+    Font* getFont(RenderingResourceIdentifier identifier) const
```
Any reason for this one to not return a `RefPtr`?

kkinnunen-apple replied:
When writing this, I was thinking if I was the reviewer, I would not notice this..

Yeah, Skia backend uses display lists incorrectly in raster threads, and Font is not threadsaferefcounted, so adding refs will crash gtk/wpe in EWS. Naturally that use is totally incorrect, but somebody already approved the code..

mattwoodrow commented on Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
```diff
@@ -438,6 +466,14 @@ void RemoteRenderingBackend::cacheGradient(Ref<Gradient>&& gradient)
+
 void RemoteRenderingBackend::cacheFilter(Ref<Filter>&& filter)
 {
     ASSERT(!RunLoop::isMain());
```
Not really part of this change, but it'd be nice for these checks to be switched to `assertIsCurrent(workQueue())` at some point.

kkinnunen-apple replied:
Yeah, I have pending patches to do something with them..

mattwoodrow commented on Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
```diff
@@ -438,6 +466,14 @@ void RemoteRenderingBackend::cacheGradient(Ref<Gradient>&& gradient)
+{
+    assertIsCurrent(workQueue());
+    bool success = m_remoteResourceCache.releaseGradient(identifier);
+    MESSAGE_CHECK(success, "Gradientreleased before being cached.");
```
Missing whitespace

mattwoodrow commented on Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
```diff
@@ -46,6 +46,10 @@
+#include <WebCore/Filter.h>
+#include <WebCore/Font.h>
+#include <WebCore/FontCustomPlatformData.h>
+#include <WebCore/Gradient.h>
```
Could/should we forward declare some of these?

kkinnunen-apple committed:
GPU process uses same map for all resources
https://bugs.webkit.org/show_bug.cgi?id=288587
rdar://145647088

Reviewed by Matt Woodrow.

DisplayList::ResourceHeap, RemoteResourceCacheProxy, RemoteResourceCache
would use std::variant to put unrelated types ImageBuffer, NativeImage,
Gradient, DecomposedGlyphs, Font, FontCustomPlatformData to the same
hash map.

These types are distinct types that are never acted upon generically,
rather each have distinct insert, remove and query locations that
use the specific type. Thus it does not have any benefit to store
these in the same map.

Performance-wise, the type is the best discriminator and thus inserts,
lookups are fastest with the smallest amount of items in the maps.

Memory-wise, due to the superlinear overallocation of hashmaps, the
smallest amount if items in each map is the best approach when maps
are of non-trivial size.

Code structure-wise, having the different types in the same map causes
the antipattern of having the identifier as RenderingResourceIdentifier
instead of type-specific identifiers. It also requires error-prone and
buggy code instance counting, downcasting and result filtering code
that is just not needed at all.

Fix by storing the instances in HashSets and HashMaps that are distinct
to each type.

Implement the maps in both RemoteResourceCache and
DisplayList::ResourceHeap. These have different use-cases and are not
used with each others interfaces. Using DisplayList::ResourceHeap as the
implementation of RemoteResourceCache would leak remote resource cache
related code to DisplayList::ResourceHeap. Also
DisplayList::ResourceHeap is unneeded functionality and will be removed
in later commits.

Later commits will improve the type safety by introducing the type
specific identifiers.

* Source/WTF/wtf/WeakHashSet.h:
Add capability to store WeakRefs that come from WeakFactory.

* Source/WebCore/platform/graphics/DecomposedGlyphs.cpp:
(WebCore::DecomposedGlyphs::~DecomposedGlyphs):
* Source/WebCore/platform/graphics/DecomposedGlyphs.h:
* Source/WebCore/platform/graphics/Gradient.cpp:
(WebCore::Gradient::~Gradient):
* Source/WebCore/platform/graphics/Gradient.h:
* Source/WebCore/platform/graphics/NativeImage.cpp:
(WebCore::NativeImage::~NativeImage):
* Source/WebCore/platform/graphics/NativeImage.h:
* Source/WebCore/platform/graphics/RenderingResource.h:
(WebCore::RenderingResource::addObserver):
(WebCore::RenderingResource::~RenderingResource): Deleted.
(WebCore::RenderingResource::removeObserver): Deleted.
Remove unused function, it makes no sense for weak observers. Instead,
invalidate the observer.

* Source/WebCore/platform/graphics/displaylists/DisplayListItem.cpp:
(WebCore::DisplayList::applyFilteredImageBufferItem):
(WebCore::DisplayList::applyImageBufferItem):
(WebCore::DisplayList::applyNativeImageItem):
(WebCore::DisplayList::applyDrawGlyphs):
(WebCore::DisplayList::applyDrawDecomposedGlyphs):
* Source/WebCore/platform/graphics/displaylists/DisplayListResourceHeap.h:
(WebCore::DisplayList::ResourceHeap::add):
(WebCore::DisplayList::ResourceHeap::getImageBuffer):
(WebCore::DisplayList::ResourceHeap::getNativeImage):
(WebCore::DisplayList::ResourceHeap::getSourceImage):
(WebCore::DisplayList::ResourceHeap::getDecomposedGlyphs const):
(WebCore::DisplayList::ResourceHeap::getGradient const):
(WebCore::DisplayList::ResourceHeap::getFilter const):
(WebCore::DisplayList::ResourceHeap::getFont const):
(WebCore::DisplayList::ResourceHeap::clearAllResources):
(WebCore::DisplayList::ResourceHeap::getFontCustomPlatformData const): Deleted.
(WebCore::DisplayList::ResourceHeap::resources const): Deleted.
(WebCore::DisplayList::ResourceHeap::removeImageBuffer): Deleted.
(WebCore::DisplayList::ResourceHeap::removeRenderingResource): Deleted.
(WebCore::DisplayList::ResourceHeap::removeFont): Deleted.
(WebCore::DisplayList::ResourceHeap::removeFontCustomPlatformData): Deleted.
(WebCore::DisplayList::ResourceHeap::clearAllImageResources): Deleted.
(WebCore::DisplayList::ResourceHeap::clearAllDrawingResources): Deleted.
(WebCore::DisplayList::ResourceHeap::get const): Deleted.
(WebCore::DisplayList::ResourceHeap::remove): Deleted.
(WebCore::DisplayList::ResourceHeap::checkInvariants const): Deleted.
(): Deleted.
* Source/WebCore/platform/graphics/filters/FilterFunction.cpp:
(WebCore::FilterFunction::~FilterFunction):
* Source/WebCore/platform/graphics/filters/FilterFunction.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.cpp:
(WebKit::RemoteRenderingBackend::releaseNativeImage):
(WebKit::RemoteRenderingBackend::releaseFont):
(WebKit::RemoteRenderingBackend::releaseFontCustomPlatformData):
(WebKit::RemoteRenderingBackend::releaseDecomposedGlyphs):
(WebKit::RemoteRenderingBackend::releaseGradient):
(WebKit::RemoteRenderingBackend::releaseFilter):
(WebKit::RemoteRenderingBackend::releaseRenderingResource): Deleted.
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.h:
* Source/WebKit/GPUProcess/graphics/RemoteRenderingBackend.messages.in:
* Source/WebKit/GPUProcess/graphics/RemoteResourceCache.cpp:
(WebKit::RemoteResourceCache::cacheNativeImage):
(WebKit::RemoteResourceCache::releaseNativeImage):
(WebKit::RemoteResourceCache::cachedNativeImage const):
(WebKit::RemoteResourceCache::cacheDecomposedGlyphs):
(WebKit::RemoteResourceCache::releaseDecomposedGlyphs):
(WebKit::RemoteResourceCache::cachedDecomposedGlyphs const):
(WebKit::RemoteResourceCache::cacheGradient):
(WebKit::RemoteResourceCache::releaseGradient):
(WebKit::RemoteResourceCache::cachedGradient const):
(WebKit::RemoteResourceCache::cacheFilter):
(WebKit::RemoteResourceCache::releaseFilter):
(WebKit::RemoteResourceCache::cachedFilter const):
(WebKit::RemoteResourceCache::cacheFont):
(WebKit::RemoteResourceCache::releaseFont):
(WebKit::RemoteResourceCache::cachedFont const):
(WebKit::RemoteResourceCache::cacheFontCustomPlatformData):
(WebKit::RemoteResourceCache::releaseFontCustomPlatformData):
(WebKit::RemoteResourceCache::cachedFontCustomPlatformData const):
(WebKit::RemoteResourceCache::releaseAllResources):
(WebKit::RemoteResourceCache::releaseAllDrawingResources):
(WebKit::RemoteResourceCache::releaseAllImageResources):
(WebKit::RemoteResourceCache::releaseRenderingResource): Deleted.
* Source/WebKit/GPUProcess/graphics/RemoteResourceCache.h:
(WebKit::RemoteResourceCache::resourceHeap const): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.cpp:
(WebKit::RemoteRenderingBackendProxy::releaseImageBuffer):
(WebKit::RemoteRenderingBackendProxy::releaseNativeImage):
(WebKit::RemoteRenderingBackendProxy::releaseFont):
(WebKit::RemoteRenderingBackendProxy::releaseFontCustomPlatformData):
(WebKit::RemoteRenderingBackendProxy::releaseDecomposedGlyphs):
(WebKit::RemoteRenderingBackendProxy::releaseGradient):
(WebKit::RemoteRenderingBackendProxy::releaseFilter):
(WebKit::RemoteRenderingBackendProxy::releaseAllDrawingResources):
(WebKit::RemoteRenderingBackendProxy::releaseAllImageResources):
(WebKit::RemoteRenderingBackendProxy::releaseRenderingResource): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteRenderingBackendProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.cpp:
(WebKit::RemoteResourceCacheProxy::~RemoteResourceCacheProxy):
(WebKit::RemoteResourceCacheProxy::clear):
(WebKit::RemoteResourceCacheProxy::recordDecomposedGlyphsUse):
(WebKit::RemoteResourceCacheProxy::recordGradientUse):
(WebKit::RemoteResourceCacheProxy::recordFilterUse):
(WebKit::RemoteResourceCacheProxy::recordNativeImageUse):
(WebKit::RemoteResourceCacheProxy::willDestroyNativeImage):
(WebKit::RemoteResourceCacheProxy::willDestroyGradient):
(WebKit::RemoteResourceCacheProxy::willDestroyDecomposedGlyphs):
(WebKit::RemoteResourceCacheProxy::willDestroyFilter):
(WebKit::RemoteResourceCacheProxy::clearRenderingResources):
(WebKit::RemoteResourceCacheProxy::clearNativeImages):
(WebKit::RemoteResourceCacheProxy::finalizeRenderingUpdateForFonts):
(WebKit::RemoteResourceCacheProxy::remoteResourceCacheWasDestroyed):
(WebKit::RemoteResourceCacheProxy::releaseMemory):
(WebKit::RemoteResourceCacheProxy::releaseAllImageResources):
(WebKit::RemoteResourceCacheProxy::imagesCount const): Deleted.
(WebKit::RemoteResourceCacheProxy::cachedNativeImage const): Deleted.
(WebKit::RemoteResourceCacheProxy::releaseRenderingResource): Deleted.
(WebKit::RemoteResourceCacheProxy::clearRenderingResourceMap): Deleted.
(WebKit::RemoteResourceCacheProxy::clearNativeImageMap): Deleted.
* Source/WebKit/WebProcess/GPU/graphics/RemoteResourceCacheProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::remoteImagesCountForTesting const):

Canonical link: https://commits.webkit.org/291617@main
